diff --git a/binding.gyp b/binding.gyp
index bb40aa7e8d41644ba690761e9064fc1d0f2dcd6b..724b177f43bf6ed7cf60add3f23317be2018f6ff 100644
--- a/binding.gyp
+++ b/binding.gyp
@@ -24,14 +24,11 @@
       "defines": [
         "NAPI_VERSION=<(napi_build_version)",
       ],
-      "cflags_cc": [
-        "-std=c++17"
-      ],
       "conditions": [
         ["OS=='mac'", {
           "xcode_settings": {
             "GCC_SYMBOLS_PRIVATE_EXTERN": "YES", # -fvisibility=hidden
-            "CLANG_CXX_LANGUAGE_STANDARD": "c++17",
+            "CLANG_CXX_LANGUAGE_STANDARD": "<(cxxstd)",
             "MACOSX_DEPLOYMENT_TARGET": "10.9",
           },
         }],
@@ -39,15 +36,17 @@
           "msvs_settings": {
             "VCCLCompilerTool": {
               "AdditionalOptions": [
-                "/std:c++17",
+                "/std:<(cxxstd)",
               ],
               "RuntimeLibrary": 0,
             },
           },
         }],
         ["OS == 'linux'", {
           "cflags_cc": [
-            "-Wno-cast-function-type"
+            "-std=<(cxxstd)",
+            "-fvisibility=hidden",
+            "-Wno-cast-function-type",
           ]
         }],
       ]
@@ -75,5 +74,6 @@
     "openssl_fips": "",
     "v8_enable_pointer_compression%": 0,
     "v8_enable_31bit_smis_on_64bit_arch%": 0,
+    "cxxstd%": "<!(node -p \"parseInt(process.env.npm_config_target ?? process.versions.node) < 22 ? 'c++17' : 'c++20'\")",
   }
 }
